pkg/storegateway/indexheader/lazy_binary_reader.go

```diff
@@ -294,6 +296,8 @@ func (r *LazyBinaryReader) unloadIfIdleSince(ts int64) error {
 	r.readerMx.Lock()
 	defer r.readerMx.Unlock()
 
+	r.readerInUse.Wait()
```

Comment:
I can't quite follow why this `WaitGroup` is needed. It seems like holding the read lock in `getOrLoadReader()` is enough to prevent a nil pointer access by the callers.

Reply:
Problem is that function like `IndexVersion` will get the reader, but when it actually tries to call `reader.IndexVersion()`, it may find that reader is now being closed, because `unloadIfIdleSince` was just called. This was caught by `TestLazyBinaryReader_LoadUnloadRaceCondition` test.